### PR TITLE
Fix Thread Safety Problem in Bulk Batching

### PR DESCRIPTION
**Pull request by johnkeith**
See [Basecamp discussion](https://3.basecamp.com/4119850/buckets/9653296/messages/2466524698) for details.

### CODE REVIEW

File: lib/elastic_record/connection.rb [hunk 1]
```diff
@@ -14,7 +14,7 @@ def initialize(servers, options = {})
       self.request_count      = 0
       self.max_request_count  = 100
       self.options            = options.symbolize_keys
-      self.bulk_actions       = nil
+      self.bulk_actions       = {}
```
Comment (ebarendt) [inline lib/elastic_record/connection.rb]: Instead of this, try

```ruby
Thread.current['bulk_actions'] = nil
```

Look at this project for other uses of `Thread.current`

**Comment by AlexMooney** [comment]
This looks very promising!  Could definitely explain ES being out of sync.

**Comment by whereisciao** [comment]
> This looks very promising! Could definitely explain ES being out of sync.

What problems were you guys seeing? Was it with the DataImport code or something else? I ask so Support can keep an eye on things.

**Comment by AlexMooney** [comment]
@whereisciao it's been a longstanding problem that ES can desync from postgres.  I can all but guarantee that something is out of sync in production, at any given time.  [Here's an example](https://data-axle.com/people?scope=9afc6431518&columns=person,emails,emails_count).

**Comment by matthuhiggins** [comment]
BTW: https://github.com/data-axle/elastic_record/pull/59

That one was reverted due to minor gotchas.

**Comment by johnkeith** [comment]
@matthuhiggins, @ebarendt - that is interesting, do y'all remember any of the gotcha scenarios?

In lieu of the problems we are seeing now, how about we roll this out and handle the corner cases as they arise?

**Comment by ebarendt** [comment]
Honeybadgers probably.

What's up with the failing tests?

**Comment by johnkeith** [comment]
@ebarendt - the tests are good now. The problem was in how we were reassigning that current_bulk_batch value here https://github.com/data-axle/elastic_record/pull/119/commits/e1ab73b2620be3cd69dcce57eecc0a1191b54f44#diff-52b112257e8727735e5a2989a0fb4673R112. I've made things more explicit and the tests are happy. 

I'm going to get this pushed to QA for further testing.

**Comment by matthuhiggins** [comment]
Curious: are you trying to run concurrency in production as a new idea?

**Comment by johnkeith** [comment]
@matthuhiggins no, this is in response to a bug we were dealing with where RecordLists were not being properly populated or truncated. This is just to try to get things working properly in Sidekiq when we have multiple workers running bulk batches.

**Comment by johnkeith** [comment]
These changes are out on QA3 for testing.

**Comment by matthuhiggins** [comment]
>  This is just to try to get things working properly in Sidekiq when we have multiple workers running bulk batches.

In the same thread?

**Comment by ebarendt** [comment]
Alex and I have seen, in development, problems where PG counts != ES counts after doing large operations. I use the sidekiq `-c 4` option usually, so I'm fairly certain this contributes to the problem.

Production-wise, not sure if this would cause record lists being out of sync since they're all `-c 1`.

**Comment by matthuhiggins** [comment]
> I use the sidekiq -c 4 option usually, so I'm fairly certain this contributes to the problem.

No question that this causes bugs, if you concurrency is greater than 1. It's good to fix this bug.

> Production-wise, not sure if this would cause record lists being out of sync since they're all -c 1.

This will not help anything in production.

**Comment by johnkeith** [comment]
@matthuhiggins - @pumpkineater brought up the same point, that Production is 1:1 in terms of workers/threads. 

It makes sense that we thought we had the bug reproduced in QA, as the worker that processed the adding/truncating jobs for RecordLists has two threads. 

Adding to and truncating RecordLists in production seems to be working without issues as I am testing now, so that means the reason for the bug with that Tier 1 list must be something else. 😞

**Comment by ebarendt** [comment]
In the past they have uploaded tier1/tier2 files with duplicate igids in them. Check that first.

**Comment by johnkeith** [comment]
Thanks @ebarendt, I'll take a look. I'm going to merge this one in, since it still fixes an issue. Then I'll bump the gem version over on content system. I also see Scrape and Transfer station are both running Sidekiq with workers that have multiple threads - maybe we should update those as well? (I'm not sure about auth time, no access to Sidekiq).

**Comment by ebarendt** [comment]
Scrape does use ES - or did at one point. Transfer Station does not.
